src/Storages/MaterializedView/RefreshTimers.cpp

```diff
@@ -0,0 +1,243 @@
+        case IntervalKind::Minute:
+            minutes = std::chrono::minutes{val};
+            break;
+        case IntervalKind::Hour:
+            hours = std::chrono::hours{val};
+            break;
+        case IntervalKind::Day:
+            days = std::chrono::days{val};
+            break;
+        case IntervalKind::Week:
+            weeks = std::chrono::weeks{val};
+            break;
+        case IntervalKind::Month:
+            months = std::chrono::months{val};
+            break;
+        case IntervalKind::Year:
+            years = std::chrono::years{val};
+            break;
+        default:
+            break;
+    }
+}
+
+RefreshEveryTimer::RefreshEveryTimer(const ASTTimePeriod & time_period, const ASTTimeInterval * time_offset)
+    : offset(time_offset)
+    , value{static_cast<UInt32>(time_period.value)}
+    , kind{time_period.kind}
+{
+    // TODO: validate invariants
+}
+
+std::chrono::sys_seconds RefreshEveryTimer::next(std::chrono::system_clock::time_point tp) const
+{
+    if (value == 0)
+        return std::chrono::floor<std::chrono::seconds>(tp);
+    switch (kind)
+    {
+        case IntervalKind::Second:
+            return alignedToSeconds(tp);
+        case IntervalKind::Minute:
+            return alignedToMinutes(tp);
+        case IntervalKind::Hour:
+            return alignedToHours(tp);
+        case IntervalKind::Day:
+            return alignedToDays(tp);
+        case IntervalKind::Week:
+            return alignedToWeeks(tp);
+        case IntervalKind::Month:
+            return alignedToMonths(tp);
+        case IntervalKind::Year:
+            return alignedToYears(tp);
+        default:
+            return std::chrono::ceil<std::chrono::seconds>(tp);
+    }
+}
+
+std::chrono::sys_seconds RefreshEveryTimer::alignedToYears(std::chrono::system_clock::time_point tp) const
+{
+    using namespace std::chrono_literals;
+
+    auto tp_days = std::chrono::floor<std::chrono::days>(tp);
+    std::chrono::year_month_day tp_ymd(tp_days);
+    auto normalize_years = [](std::chrono::year year) -> std::chrono::sys_days
+    {
+        return year / std::chrono::January / 1d;
+    };
+
+    auto prev_years = normalize_years(tp_ymd.year());
+    if (auto prev_time = offset.after(prev_years); prev_time > tp)
+        return prev_time;
+
+    auto next_years = normalize_years(tp_ymd.year() + std::chrono::years{1});
+    return offset.after(next_years);
+}
+
+std::chrono::sys_seconds RefreshEveryTimer::alignedToMonths(std::chrono::system_clock::time_point tp) const
+{
+    using namespace std::chrono_literals;
+
+    auto tp_days = std::chrono::floor<std::chrono::days>(tp);
+    std::chrono::year_month_day tp_ymd(tp_days);
+    auto normalize_months = [](const std::chrono::year_month_day & ymd, unsigned month_value) -> std::chrono::sys_days
+    {
+        return ymd.year() / std::chrono::month{month_value} / 1d;
+    };
+
+    auto prev_month_value = static_cast<unsigned>(tp_ymd.month()) / value * value;
```

Comment:
I assume it's like EVERY N'th month and aligning it month to year, second to minute, etc. Probably we can extend syntax by adding alignment time to statement like `EVERY N'th MONTH SINCE 2023`, or `EVERY 13 SECOND SINCE '2013-01-13 13:13:13'`